docs/theories/categories.md

```diff
@@ -0,0 +1,78 @@
+# Maths in Lean : category theory
+
+The `category` typeclass is defined in [category_theory/category.lean](https://github.com/leanprover/mathlib/blob/master/category_theory/category.lean).
+It depends on the type of the objects, so for example we might write `category (Type u)` if we're talking about a category whose objects are types (in universe `u`).
+Some care is needed with universes (see the section [Universes](##markdown-header-universes)), and end users may often prefer the abbreviations `small_category` and `large_category`.
+
+Functors (which are a structure, not a typeclass) are defined in [category_theory/functor/default.lean](https://github.com/leanprover/mathlib/blob/master/category_theory/functor/default.lean),
+along with identity functors and functor composition.
+
+Natural transformations, and their compositions, are defined in [category_theory/natural_transformation.lean](https://github.com/leanprover/mathlib/blob/master/category_theory/natural_transformation.lean).
+
+## Universes
+
+Unfortunately in a category theory library we have to deal with universes carefully. We have the following:
+
+````
+category.{u₁ v₁}          : Type (max (u₁+1) (v₁+1))
+C                         : category.{u₁ v₁}
+D                         : category.{u₂ v₂}
+Functor C D               : Type (max u₁ u₂ v₁ v₂)
+F G                       : Functor C D
+NaturalTransformation F G : Type (max u₁ v₂)
+FunctorCategory C D       : category.{(max u₁ u₂ v₁ v₂) (max u₁ v₂)}
+````
+
+Note then that if we specialise to small categories, where `uᵢ = vᵢ`, then `FunctorCategory C D : category.{(max u₁ u₂) (max u₁ u₂)}`, and so is again a small category.
+If `C` is a small category and `D` is a large category (i.e. `u₂ = v₂+1`), and `v₂ = v₁` then we have `FunctorCategory C D : category.{v₁+1 v₁}` so is again a large category.
+
+Whenever you want to write code uniformly for small and large categories (which you do by talking about categories whose universe levels `u` and `v` are unrelated), you will find that
+Lean's `variable` mechanism doesn't always work, and the following trick is often helpful:
+
+````
+variables {C : Type u₁} [𝒞 : category.{u₁ v₁} C]
+variables {D : Type u₂} [𝒟 : category.{u₂ v₂} D]
+include 𝒞 𝒟
+````
+
+
+## Notation
+
+### Categories
+
+We use the `⟶` (`\hom`) arrow to denote sets of morphisms, as in `X ⟶ Y`.
+This leaves the actual category implicit; it is inferred from the type of X and Y but typeclass inference.
+
+We use `𝟙` (`\b1`) to denote identity morphisms, as in `𝟙 X`.
+
+We use `≫` (`\gg`) to denote composition of morphisms, as in `f ≫ g`, which means "`f` followed by `g`".
+This is the opposite order than the usual convention (which is lame).  
```

Comment:
Mathematicians are going to love you for this [/sarcasm]. Is it possible to also have notation for "ordinary" composition? Would it even be possible to overload `\circ`? Or will madness ensue? If so, then preferably something that reminds of `\circ`.

Reply:
I will investigate overloading `\circ`, or finding something nearby.

I'd been hoping that everyone would be relieved that starting to write mathematics in Lean gave them a good opportunity to abandon the no-good-terrible-horrible `\circ` notation after so many years. :-)

Reply:
Regardless of one's personal preferences, for the purpose of formalizing existing mathematics, it's greatly helpful if the Lean notation is as similar as possible to the source notation; and that will almost always require having some form of applicative-order composition notation available.

In my own project, I just write ``local notation f ` ∘ `:80 g:80 := g ≫ f`` at the top of most files, and this works pretty well for me, though I occasionally need to write `function.comp` for the usual `∘` operator. Not a very good solution for the library, though.